File: 2375. Construct Smallest Number From DI String/main.py
# total sam :)

class Solution:
    def smallestNumber(self, pattern: str) -> str:
        used = set()
        cur = []

        def backtrack(i):
            if i == len(pattern):
                return True
            if pattern[i] == "I":
                for j in range(cur[-1] + 1, 10):
                    if j in used: continue
                    cur.append(j)
                    used.add(j)
                    if backtrack(i + 1):
                        return True
                    cur.pop()
                    used.remove(j)
            else:
                for j in range(1, cur[-1]):
                    if j in used: continue
                    cur.append(j)
                    used.add(j)
                    if backtrack(i + 1):
                        return True
                    cur.pop()
                    used.remove(j)
            return False
            
        for i in range(1, 10):
            cur.append(i)
            used.add(i)
            if backtrack(0):
                return "".join([str(n) for n in cur])
            used.remove(i)
            cur.pop()

File: 2375. Construct Smallest Number From DI String/main.go
package main

import (
	"strconv"
	"strings"
)

func smallestNumber(pattern string) string {
	stack, res := make([]int, 0), make([]string, len(pattern)+1)

	for i := 0; i < len(pattern)+1; i++ {
		stack = append(stack, i+1)

		for len(stack) > 0 && (i == len(pattern) || pattern[i] == 'I') {
			res = append(res, strconv.Itoa(stack[len(stack)-1]))
			stack = stack[:len(stack)-1]
		}
	}

	return strings.Join(res, "")
}
